File: debop4s-timeperiod/src/test/scala/com/github/debop4s/timeperiod/tests/timeranges/QuarterRangeCollectionTest.scala
package com.github.debop4s.timeperiod.tests.timeranges


import com.github.debop4s.timeperiod.Quarter
import com.github.debop4s.timeperiod.tests.AbstractTimePeriodTest
import com.github.debop4s.timeperiod.timerange.{QuarterRange, QuarterRangeCollection}
import com.github.debop4s.timeperiod.utils.Times._

/**
 * QuarterRangeCollectionTest
 * Created by debop on 2014. 2. 16.
 */
class QuarterRangeCollectionTest extends AbstractTimePeriodTest {

    test("year test") {
        val moment = asDate(2009, 2, 15)
        val year = yearOf(moment.getYear, moment.getMonthOfYear)
        val qr = QuarterRangeCollection(moment, 3)

        qr.startYear should equal(year)
        qr.start should equal(asDate(year, 1, 1))
    }

    test("single quarter") {
        val startYear = 2004
        val startQuarter = Quarter.Second

        val qr = QuarterRangeCollection(startYear, startQuarter, 1)

        qr.quarterCount should equal(1)
        qr.startYear should equal(startYear)
        qr.startQuarter should equal(startQuarter)
        qr.endYear should equal(startYear)
        qr.endQuarter should equal(startQuarter)

        val quarters = qr.getQuarters
        quarters.size should equal(1)
        quarters(0).isSamePeriod(QuarterRange(startYear, startQuarter)) should equal(true)
    }

    test("first calendar halfyears") {
        val startYear = 2004
        val startQuarter = Quarter.First
        val quarterCount = 5

        val qrs = QuarterRangeCollection(startYear, startQuarter, quarterCount)

        qrs.quarterCount should equal(quarterCount)
        qrs.startYear should equal(startYear)
        qrs.startQuarter should equal(startQuarter)
        qrs.endYear should equal(startYear + 1)
        qrs.endQuarter should equal(Quarter.First)

        val quarters = qrs.getQuarters

        quarters.size should equal(quarterCount)
        quarters(0).isSamePeriod(QuarterRange(startYear, Quarter.First)) should equal(true)
        quarters(1).isSamePeriod(QuarterRange(startYear, Quarter.Second)) should equal(true)
        quarters(2).isSamePeriod(QuarterRange(startYear, Quarter.Third)) should equal(true)
        quarters(3).isSamePeriod(QuarterRange(startYear, Quarter.Fourth)) should equal(true)
        quarters(4).isSamePeriod(QuarterRange(startYear + 1, Quarter.First)) should equal(true)
    }

    test("second calendar halfyears") {
        val startYear = 2004
        val startQuarter = Quarter.Second
        val quarterCount = 5

        val qrs = QuarterRangeCollection(startYear, startQuarter, quarterCount)

        qrs.quarterCount should equal(quarterCount)
        qrs.startYear should equal(startYear)
        qrs.startQuarter should equal(startQuarter)
        qrs.endYear should equal(startYear + 1)
        qrs.endQuarter should equal(Quarter.Second)

        val quarters = qrs.getQuarters

        quarters.size should equal(quarterCount)
        quarters(0).isSamePeriod(QuarterRange(startYear, Quarter.Second)) should equal(true)
        quarters(1).isSamePeriod(QuarterRange(startYear, Quarter.Third)) should equal(true)
        quarters(2).isSamePeriod(QuarterRange(startYear, Quarter.Fourth)) should equal(true)
        quarters(3).isSamePeriod(QuarterRange(startYear + 1, Quarter.First)) should equal(true)
        quarters(4).isSamePeriod(QuarterRange(startYear + 1, Quarter.Second)) should equal(true)
    }

}

File: debop4s-timeperiod/src/test/java/debop4s/timeperiod/tests/timeranges/QuarterRangeCollectionTest.java
package debop4s.timeperiod.tests.timeranges;

import debop4s.timeperiod.Quarter;
import debop4s.timeperiod.tests.TimePeriodTestBase;
import debop4s.timeperiod.timerange.QuarterRange;
import debop4s.timeperiod.timerange.QuarterRangeCollection;
import debop4s.timeperiod.utils.Times;
import lombok.extern.slf4j.Slf4j;
import org.joda.time.DateTime;
import org.junit.Test;

import java.util.List;

import static org.fest.assertions.Assertions.assertThat;

/**
 * debop4s.timeperiod.test.timeranges.QuarterRangeCollectionFunSuite
 *
 * @author 배성혁 sunghyouk.bae@gmail.com
 * @since 13. 5. 25. 오후 11:18
 */
@Slf4j
public class QuarterRangeCollectionTest extends TimePeriodTestBase {

    @Test
    public void yearBaseMonthTest() {

        DateTime moment = Times.asDate(2009, 2, 15);
        int year = Times.yearOf(moment.getYear(), moment.getMonthOfYear());
        QuarterRangeCollection quarterRanges = new QuarterRangeCollection(moment, 3);

        assertThat(quarterRanges.getStart()).isEqualTo(Times.asDate(year, 1, 1));
    }

    @Test
    public void singleQuarterTest() {
        final int startYear = 2004;
        final Quarter startQuarter = Quarter.Second;

        QuarterRangeCollection quarterRanges = new QuarterRangeCollection(startYear, startQuarter, 1);

        assertThat(quarterRanges.getQuarterCount()).isEqualTo(1);
        assertThat(quarterRanges.getStartQuarter()).isEqualTo(startQuarter);
        assertThat(quarterRanges.getEndYear()).isEqualTo(startYear);
        assertThat(quarterRanges.getEndQuarter()).isEqualTo(startQuarter);

        List<QuarterRange> quarters = quarterRanges.getQuarters();
        assertThat(quarters.size()).isEqualTo(1);
        assertThat(quarters.get(0).isSamePeriod(new QuarterRange(2004, Quarter.Second))).isTrue();
    }

    @Test
    public void firstCalendarHalfyears() {

        final int startYear = 2004;
        final Quarter startQuarter = Quarter.First;
        final int quarterCount = 5;

        QuarterRangeCollection quarterRanges = new QuarterRangeCollection(startYear, startQuarter, quarterCount);

        assertThat(quarterRanges.getQuarterCount()).isEqualTo(quarterCount);
        assertThat(quarterRanges.getStartQuarter()).isEqualTo(startQuarter);
        assertThat(quarterRanges.getEndYear()).isEqualTo(startYear + 1);
        assertThat(quarterRanges.getEndQuarter()).isEqualTo(Quarter.First);

        List<QuarterRange> quarters = quarterRanges.getQuarters();

        assertThat(quarters.size()).isEqualTo(quarterCount);
        assertThat(quarters.get(0).isSamePeriod(new QuarterRange(2004, Quarter.First))).isTrue();
        assertThat(quarters.get(1).isSamePeriod(new QuarterRange(2004, Quarter.Second))).isTrue();
        assertThat(quarters.get(2).isSamePeriod(new QuarterRange(2004, Quarter.Third))).isTrue();
        assertThat(quarters.get(3).isSamePeriod(new QuarterRange(2004, Quarter.Fourth))).isTrue();
        assertThat(quarters.get(4).isSamePeriod(new QuarterRange(2005, Quarter.First))).isTrue();
    }

    @Test
    public void secondCalendarHalfyears() {
        final int startYear = 2004;
        final Quarter startQuarter = Quarter.Second;
        final int quarterCount = 5;

        QuarterRangeCollection quarterRanges = new QuarterRangeCollection(startYear, startQuarter, quarterCount);

        assertThat(quarterRanges.getQuarterCount()).isEqualTo(quarterCount);
        assertThat(quarterRanges.getStartQuarter()).isEqualTo(startQuarter);
        assertThat(quarterRanges.getEndYear()).isEqualTo(startYear + 1);
        assertThat(quarterRanges.getEndQuarter()).isEqualTo(Quarter.Second);

        List<QuarterRange> quarters = quarterRanges.getQuarters();

        assertThat(quarters.size()).isEqualTo(quarterCount);
        assertThat(quarters.get(0).isSamePeriod(new QuarterRange(2004, Quarter.Second))).isTrue();
        assertThat(quarters.get(1).isSamePeriod(new QuarterRange(2004, Quarter.Third))).isTrue();
        assertThat(quarters.get(2).isSamePeriod(new QuarterRange(2004, Quarter.Fourth))).isTrue();
        assertThat(quarters.get(3).isSamePeriod(new QuarterRange(2005, Quarter.First))).isTrue();
        assertThat(quarters.get(4).isSamePeriod(new QuarterRange(2005, Quarter.Second))).isTrue();
    }
}
